File: dotnet/src/MableBank.Application/RecordingMetrics.cs
using MableBank.Application.Telemetry;

namespace MableBank.Application;

public sealed class RecordingMetrics : IMetrics
{
    public IList<TransferProcessedTelemetry> Records { get; } = new List<TransferProcessedTelemetry>();

    public IList<MetricIncrement> Increments { get; } = new List<MetricIncrement>();

    public void RecordTransferProcessed(TransferProcessedTelemetry telemetry)
    {
        ArgumentNullException.ThrowIfNull(telemetry);
        Records.Add(telemetry);
        Increments.Add(telemetry.ToMetricIncrement());
    }
}

File: nodejs-typescript/src/application/RecordingMetrics.ts
import type { Metrics } from './Metrics.js';
import type {
  MetricIncrement,
  TransferProcessedTelemetry,
} from './telemetry/TransferProcessedTelemetry.js';

export class RecordingMetrics implements Metrics {
  readonly records: TransferProcessedTelemetry[] = [];
  readonly increments: MetricIncrement[] = [];

  recordTransferProcessed(telemetry: TransferProcessedTelemetry): void {
    this.records.push(telemetry);
    this.increments.push(telemetry.toMetricIncrement());
  }
}

export type { MetricIncrement };
